runtime-light/stdlib/math/random-functions.h

```diff
@@ -42,3 +50,92 @@ inline int64_t f$rand(int64_t l, int64_t r) noexcept {
+inline constexpr int64_t a1 = 53668;
+inline constexpr int64_t b1 = 40014;
+inline constexpr int64_t c1 = 12211;
+
+inline constexpr int64_t lcg_prime2 = (int64_t(1) << 31) - int64_t(249);
+inline constexpr int64_t a2 = 52774;
+inline constexpr int64_t b2 = 40692;
+inline constexpr int64_t c2 = 3791;
+
+// invariant:   |lcg_modmult(S1)|  \in  (-P1 .. P1)
+static_assert(b1 * (a1 - 1) - c1 * (0) < lcg_prime1);                 // max lcg_modmult(S1) < P1
+static_assert(b1 * 0 - c1 * ((lcg_prime1 - 1) / a1) > (-lcg_prime1)); // min lcg_modmult(S1) > -P1
+
+// invariant:   |lcg_modmult(S2)|  \in  (-P2 .. P2)
+static_assert(b2 * (a2 - 1) - c2 * (0) < lcg_prime2);                 // max lcg_modmult(S2) < P2
+static_assert(b2 * 0 - c2 * ((lcg_prime2 - 1) / a2) > (-lcg_prime2)); // min lcg_modmult(S2) > -P2
+
+inline constexpr double lcg_value_coef = 4.656613e-10;
+
+inline int64_t lcg_modmult(int64_t a, int64_t b, int64_t c, int64_t m, int64_t s) noexcept {
+  int64_t q{s / a};
+  int64_t res{b * (s - a * q) - c * q}; // res := b * (s % a) - c * (s / a)
+  if (res < 0) {
+    res += m;
+  }
+  return res;
+}
+
+} // namespace random_impl_
+
+inline double f$lcg_value() noexcept {
+  auto& random_instance_state{RandomInstanceState::get()};
+  if (!random_instance_state.lcg_initialized) {
+    uint32_t tmp{};
+    k2::os_rnd(sizeof(tmp), std::addressof(tmp));
+    random_instance_state.lcg1 = static_cast<int64_t>(tmp) % random_impl_::lcg_prime1;
+    k2::os_rnd(sizeof(tmp), std::addressof(tmp));
+    random_instance_state.lcg2 = static_cast<int64_t>(tmp) % random_impl_::lcg_prime2;
+    random_instance_state.lcg_initialized = true;
+  }
+
+  random_instance_state.lcg1 =
+      random_impl_::lcg_modmult(random_impl_::a1, random_impl_::b1, random_impl_::c1, random_impl_::lcg_prime1, random_instance_state.lcg1);
+  random_instance_state.lcg2 =
+      random_impl_::lcg_modmult(random_impl_::a2, random_impl_::b2, random_impl_::c2, random_impl_::lcg_prime2, random_instance_state.lcg2);
+
+  int64_t z{random_instance_state.lcg1 - random_instance_state.lcg2};
+  if (z < 1) {
+    z += std::max(random_impl_::lcg_prime1, random_impl_::lcg_prime2) - 1;
+  }
+
+  return static_cast<double>(z) * random_impl_::lcg_value_coef;
+}
+
+inline kphp::coro::task<string> f$uniqid(string prefix, bool more_entropy) noexcept {
+  if (!more_entropy) {
+    co_await f$usleep(1);
+  }
+
+  auto [sec, susec]{system_seconds_and_micros()};
+  auto sec_cnt{static_cast<int32_t>(sec.count() & 0xFFFFFFFF)};  // because we'll use only 8 hex digits
+  auto susec_cnt{static_cast<int32_t>(susec.count() & 0xFFFFF)}; // because we'll use only 5 hex digits
+  constexpr size_t buf_size = 30;
+  std::array<char, buf_size> buf{};
+  // TODO is it correct to move this write to static_SB before f$usleep or not?
+  // Are we allowed to store data in static_SB before co_await and read it after?
+  RuntimeContext::get().static_SB.clean() << prefix;
```

Comment:
Let's have a reference to `RuntimeContext` as a local variable. Thus we will reduce the number of `k2::instance_state` calls

Reply:
aliased